### PR TITLE
role staking app

### PR DESCRIPTION
**Pull request by mnaamani**
Role Staking.

Requires the actors runtime module currently in [development](https://github.com/Joystream/substrate-runtime-joystream/pull/6)

Ready for review.

### CODE REVIEW

File: packages/joy-actors/src/types.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  role: Role,
+  account: AccountId,
+  joined_at: BlockNumber
+}
+
+export type Request = [AccountId, MemberId, Role, BlockNumber];
+export type Requests = Array<Request>;
+
+
+export class RoleParameters extends Struct {
```
Comment (mnaamani) [inline packages/joy-actors/src/types.ts]: see here for how to improve type definition: https://github.com/dappforce/blogstrate-ui/blob/blog/packages/joy-blogs/src/types.ts

File: packages/joy-actors/src/Actions/MyActionList.tsx [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+)(Action);
+
+type WithMyMemberIdProps = MyAccountProps & {
+    memberIdByAccountId?: Option<MemberId>,
+    requests: Array<Request>,
+};
+
+function WithMyMemberIdInner (props: WithMyMemberIdProps) {
+    if (props.memberIdByAccountId) {
+        const memberId = props.memberIdByAccountId.unwrapOr(undefined);
```
Comment (siman) [inline packages/joy-actors/src/Actions/MyActionList.tsx]: It is not recommended using `.unwrapOr(undefined)`. Almost never.
`Option` was created with an idea to give a flexible control over the situation when something doesn't exist so you can manage it w/o `try/catch`. Better use `if (opt.isSome) ...` or `if (opt.isNone) ...`

Comment (siman) [reply packages/joy-actors/src/Actions/MyActionList.tsx]: In this case if `memberIdByAccountId.isNone` you could show for example an error `"Account is required to perform this action"`.

File: packages/joy-actors/src/Actions/index.tsx [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import React from 'react'
+import { BareProps } from '@polkadot/ui-app/types';
+import { ComponentProps } from '../props';
+import MyActionList from './MyActionList';
+
+type Props = BareProps & ComponentProps;
+
+export default class Actions extends React.PureComponent<Props> {
```
Comment (siman) [inline packages/joy-actors/src/Actions/index.tsx]: Why do you need this component? It's just an empty wrapper around `MyActionList ` that passed props through. Looks redundant to me :)

File: packages/joy-actors/src/ActorsList/index.tsx [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    render() {
+        const actors: Array<string> = this.props.actors;
+
+        return (
+            <Table>
+                <Table.Header>
+                    <Table.Row>
+                        <Table.HeaderCell>Role</Table.HeaderCell>
+                        <Table.HeaderCell>Actor Account</Table.HeaderCell>
+                        <Table.HeaderCell>MemberId</Table.HeaderCell>
```
Comment (siman) [inline packages/joy-actors/src/ActorsList/index.tsx]: MemberId -> Member Id
```suggestion
                        <Table.HeaderCell>Member Id</Table.HeaderCell>
```

File: packages/joy-actors/src/ActorsList/index.tsx [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                </Table.Body>
+            </Table>
+        )
+    }
+}
+
+
+type ActorProps = BareProps & {
+    account: string,
+    actor?: Option<any>, // Option<Actor>
```
Comment (siman) [inline packages/joy-actors/src/ActorsList/index.tsx]: yeah.. to be able to use `Option<Actor>` an `Actor` should be a class that extends `Struct`. You cannot have `Option<_a_type_here_>`

File: packages/joy-actors/src/ActorsList/index.tsx [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+class ActorInner extends React.PureComponent<ActorProps> {
+    render() {
+        if (!this.props.actor) return null;
```
Comment (siman) [inline packages/joy-actors/src/ActorsList/index.tsx]:
```suggestion
        const { actor: actorOpt } = this.props;
        if (!actorOpt || actorOpt.isNone) return null;
        const actor = actorOpt.unwrap();
```

File: packages/joy-actors/src/AvailableRoles/index.tsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+import Section from '@polkadot/joy-utils/Section';
+
+import BN from 'bn.js';
+
+type Props = BareProps & ComponentProps;
+
+export default class AvailableRoles extends React.PureComponent<Props> {
+    render() {
+        return (
+            <div> {this.props.roles.map((role) => <div key={role.toString()}><RoleDisplay role={role}></RoleDisplay></div>) } </div>
```
Comment (siman) [inline packages/joy-actors/src/AvailableRoles/index.tsx]: Let me suggest a little formatting here (easier to read):
```suggestion
            <div>{this.props.roles.map((role) =>
                  <div key={role.toString()}><RoleDisplay role={role} /></div>
            )}</div>
```

File: packages/joy-actors/src/AvailableRoles/index.tsx [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    role: Role,
+    roleParams?: Option<RoleParameters>
+}
+
+class RoleDisplayInner extends React.PureComponent<RoleProps> {
+    render() {
+        const {role, roleParams} = this.props;
+        if (!roleParams) return null;
+
+        const params = roleParams.unwrapOr(undefined);
```
Comment (siman) [inline packages/joy-actors/src/AvailableRoles/index.tsx]: the same situation with [ `opt.isNone` + `opt.unwrap()` ] vs `opt.unwrapOr(undefined)`

File: packages/joy-actors/src/AvailableRoles/index.tsx [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        if (!params) return null; // no role parameters defined for this role
+
+        return (
+            <Section title={`${role.toString()}`}>
```
Comment (siman) [inline packages/joy-actors/src/AvailableRoles/index.tsx]: You want `${value}` if there should be something around that value in a string, e.g. `My age is ${age} years old`
```suggestion
            <Section title={role.toString()}>
```

File: packages/joy-actors/src/index.tsx [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+type Props = AppProps & ApiProps & I18nProps & {
+  requests?: Array<Request>,
+  actors?: Array<AccountId>,
+  roles?: Array<Role>,
+  allAccounts?: SubjectInfo,
+};
+
+type State = {
+  tabs: Array<TabItem>,
```
Comment (siman) [inline packages/joy-actors/src/index.tsx]: just small suggestion: `Array<x>` is the same as `x[]`. When `x[]` is shorter to type :)

File: packages/joy-actors/package.json [hunk 11]
```diff
@@ -0,0 +1,15 @@
+{
+  "name": "@polkadot/joy-actors",
+  "version": "0.0.1",
+  "description": "Staked roles module for Joystream node",
```
[the remark this answers is not in the document]
Comment (mnaamani) [reply packages/joy-actors/package.json]: I think you might be right about the name.

File: packages/joy-actors/src/index.tsx [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+        role
+      ),
+    } as State;
+  }
+
+  render () {
+    const { allAccounts } = this.props;
+    const { tabs } = this.state;
+    const { basePath } = this.props;
+    const hasAccounts = allAccounts && Object.keys(allAccounts).length !== 0;
```
[the remark this answers is not in the document]
Comment (mnaamani) [reply packages/joy-actors/src/index.tsx]: I guess we could even go a step further and just simply:
`const hasAccounts = allAccounts && Object.keys(allAccounts).length;`

File: packages/joy-actors/src/Actions/MyActionList.tsx (hunk 2 above)
Comment (mnaamani) [reply packages/joy-actors/src/Actions/MyActionList.tsx]: Great point. I copied this snippet from `joy-utils` and didn't really give it much thought.

**Comment by mnaamani** [comment]
@siman, I will update roles runtime module based on your review, so it will need some changes in this PR also. Don't merge this just yet.

File: packages/apps/src/routing/joy-roles.ts [hunk 13]
```diff
@@ -0,0 +1,18 @@
+import { Routes } from '../types';
+
+import Actors from '@polkadot/joy-roles/index';
```
Comment (siman) [inline packages/apps/src/routing/joy-roles.ts]:
```suggestion
import Roles from '@polkadot/joy-roles/index';
```

File: packages/joy-roles/src/types.ts [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+    const typeRegistry = getTypeRegistry();
+    typeRegistry.register({
+      Role,
+      RoleParameters,
+      Request: '(AccountId, MemberId, Role, BlockNumber)',
+      Requests: 'Vec<Request>',
+      Actor,
+    });
+  } catch (err) {
+    console.error('Failed to register custom types of actors module', err);
```
Comment (siman) [inline packages/joy-roles/src/types.ts]:
```suggestion
    console.error('Failed to register custom types of roles module', err);
```

File: packages/joy-roles/src/MyRequests/index.tsx [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+            return <div>No requests for member id: {memberId.toString()}</div>
+        }
+    }
+
+    private renderActions(requests: Array<Request>) {
+        return (
+            <Table>
+                <Table.Body>
+                    {
+                        requests.map((request: Request) => {
```
Comment (siman) [inline packages/joy-roles/src/MyRequests/index.tsx]:
```suggestion
                        requests.map(([account, _, role]: Request) => {
```

**Comment by mnaamani** [comment]
Thanks for review @siman pushed update with suggestions.

**Pull request merged**